Additional JavaScript API Token Functions

ColdAnkles opened the pull request:
### Identify the Bug or Feature request
closes  #4322

### Description of the Change
Adds additional functions to the JSAPI Token Object to streamline accessing token States and PC/NPC status.


### Possible Drawbacks
None

### Documentation Notes
Adds functions for getting and setting states, and getting all currently set states. Getting and setting PC/NPC type.

### Release Notes
Adds additional token functions to JSAPI Token.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/4323)
<!-- Reviewable:end -->

cwisniew committed:
apply compat fix for json.path.set and json.path.delete as well

ColdAnkles committed:
Added additional functions to JSAPI Token

ColdAnkles committed:
Added additional functions to JSAPI Token

ColdAnkles committed:
Client<->Server communications and trusted context tests.

ColdAnkles committed:
Client<->Server communications and trusted context tests.

thelsing commented on src/main/java/net/rptools/maptool/client/script/javascript/api/JSAPIToken.java:
```diff
@@ -199,4 +200,74 @@ public void setMap(Zone m) {
+  public void setState(String stateName, boolean aValue) {
+    boolean trusted = JSScriptEngine.inTrustedContext();
+    String playerId = MapTool.getPlayer().getName();
+    if (trusted || token.isOwner(playerId)) {
```
Does this work for token that are owned by "All players" too? I think there were issues that you have to check against this separately.

ColdAnkles replied:
I did some testing and couldn't find an issue. When a token was set to owned by "All players" a server or client GM always succeeds at the isOwner check, and when the client is a player, not GM it succeeded as well.

I could've missed something though.

thelsing replied:
If you tested the the case everthing is fine.

thelsing commented on src/main/java/net/rptools/maptool/client/script/javascript/api/JSAPIToken.java:
```diff
@@ -199,4 +200,74 @@ public void setMap(Zone m) {
+  public void setAllStates(boolean aValue) {
+    boolean trusted = JSScriptEngine.inTrustedContext();
+    String playerId = MapTool.getPlayer().getName();
+    if (trusted || token.isOwner(playerId)) {
```
See above.

thelsing commented on src/main/java/net/rptools/maptool/client/script/javascript/api/JSAPIToken.java:
```diff
@@ -199,4 +200,74 @@ public void setMap(Zone m) {
+  public void setPC() {
+    boolean trusted = JSScriptEngine.inTrustedContext();
+    String playerId = MapTool.getPlayer().getName();
+    if (trusted || token.isOwner(playerId)) {
```
See above.

thelsing commented on src/main/java/net/rptools/maptool/client/script/javascript/api/JSAPIToken.java:
```diff
@@ -199,4 +200,74 @@ public void setMap(Zone m) {
+  public void setNPC() {
+    boolean trusted = JSScriptEngine.inTrustedContext();
+    String playerId = MapTool.getPlayer().getName();
+    if (trusted || token.isOwner(playerId)) {
```
Here too.

thelsing commented on src/main/java/net/rptools/maptool/client/script/javascript/api/JSAPIToken.java:
```diff
@@ -199,4 +200,74 @@ public void setMap(Zone m) {
+
+  @HostAccess.Export
+  public String getType() {
+    if (this.token.getType() == Token.Type.PC) {
```
Wouldn't it be better to use something like the name() method of the enum? That way it would work if we ever get other token types.

thelsing commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -1738,6 +1738,18 @@ public Object getState(String property) {
     return state.get(property);
   }
 
+  public List<String> getSetStates() {
```
For me the method name is a bit confusing because getSet. Maybe getActiveStates or something similar would be better?

ColdAnkles replied:
I wasn't totally happy with it either, but wasn't coming up with anything better at the time.

ColdAnkles committed:
Merge remote-tracking branch 'origin/additional-token-functions' into additional-token-functions

ColdAnkles committed:
Clarify function name. Improved getType return.

thelsing reviewed (approved):
Github refuses to merge show me the last changed you made in the PR file view. But the commits are fine.

cwisniew committed:
Display installation directory in dev options

cwisniew committed:
Change to install location

cwisniew committed:
fix

kwvanderlinde committed:
Disable developer options by default

Also store developer options in a separate preference node to keep them separate from regular application
preferences. This should make testing and fixups a little easier.

cwisniew committed:
Move location of install directory preferences info

ColdAnkles committed:
Merge remote-tracking branch 'origin/develop' into develop
